src/bech32.h

```diff
@@ -19,11 +19,29 @@
 namespace bech32
 {
 
-/** Encode a Bech32 string. If hrp contains uppercase characters, this will cause an assertion error. */
-std::string Encode(const std::string& hrp, const std::vector<uint8_t>& values);
+enum class Encoding {
+    INVALID, //!< Failed decoding
 
-/** Decode a Bech32 string. Returns (hrp, data). Empty hrp means failure. */
-std::pair<std::string, std::vector<uint8_t>> Decode(const std::string& str);
+    BECH32,  //!< Bech32 encoding as defined in BIP173
+    BECH32M, //!< Bech32m encoding as defined in BIP350
+};
+
+/** Encode a Bech32 or Bech32m string. If hrp contains uppercase characters, this will cause an
+ *  assertion error. Encoding must be one of BECH32 or BECH32M. */
+std::string Encode(Encoding encoding, const std::string& hrp, const std::vector<uint8_t>& values);
+
+struct DecodeResult
+{
+    Encoding encoding;         //!< What encoding was detected in the result; Encoding::INVALID if failed.
+    std::string hrp;           //!< The human readable part
+    std::vector<uint8_t> data; //!< The payload (excluding checksum)
+
+    DecodeResult() : encoding(Encoding::INVALID) {}
+    DecodeResult(Encoding enc, std::string&& h, std::vector<uint8_t>&& d) : encoding(enc), hrp(std::move(h)), data(std::move(d)) {}
+};
+
+/** Decode a Bech32 string. */
+DecodeResult Decode(const std::string& str);
```

Comment:
The caller can't do anything with a `DecodeResult` where encoding is `Encoding::INVALID`, so changing this return type to `std::optional<DecodeResult>` and removing the `Encoding::INVALID` enum type feels more natural to me.

<details>
<summary>Diff</summary>

```diff
diff --git a/src/bech32.cpp b/src/bech32.cpp
index 80794ec4f1..2fd84b395b 100644
--- a/src/bech32.cpp
+++ b/src/bech32.cpp
@@ -6,6 +6,7 @@
 #include <util/vector.h>
 
 #include <assert.h>
+#include <optional>
 
 namespace bech32
 {
@@ -120,7 +121,7 @@ data ExpandHRP(const std::string& hrp)
 }
 
 /** Verify a checksum. */
-Encoding VerifyChecksum(const std::string& hrp, const data& values)
+std::optional<Encoding> VerifyChecksum(const std::string& hrp, const data& values)
 {
     // PolyMod computes what value to xor into the final values to make the checksum 0. However,
     // if we required that the checksum was 0, it would be the case that appending a 0 to a valid
@@ -129,7 +130,7 @@ Encoding VerifyChecksum(const std::string& hrp, const data& values)
     uint32_t check = PolyMod(Cat(ExpandHRP(hrp), values));
     if (check == EncodingConstant(Encoding::BECH32)) return Encoding::BECH32;
     if (check == EncodingConstant(Encoding::BECH32M)) return Encoding::BECH32M;
-    return Encoding::INVALID;
+    return {};
 }
 
 /** Create a checksum. */
@@ -165,7 +166,7 @@ std::string Encode(Encoding encoding, const std::string& hrp, const data& values
 }
 
 /** Decode a Bech32 or Bech32m string. */
-DecodeResult Decode(const std::string& str) {
+std::optional<DecodeResult> Decode(const std::string& str) {
     bool lower = false, upper = false;
     for (size_t i = 0; i < str.size(); ++i) {
         unsigned char c = str[i];
@@ -192,9 +193,9 @@ DecodeResult Decode(const std::string& str) {
     for (size_t i = 0; i < pos; ++i) {
         hrp += LowerCase(str[i]);
     }
-    Encoding result = VerifyChecksum(hrp, values);
-    if (result == Encoding::INVALID) return {};
-    return {result, std::move(hrp), data(values.begin(), values.end() - 6)};
+    auto result = VerifyChecksum(hrp, values);
+    if (!result) return {};
+    return {{*result, std::move(hrp), data(values.begin(), values.end() - 6)}};
 }
 
 } // namespace bech32
diff --git a/src/bech32.h b/src/bech32.h
index 3679ea8ccb..f2f294bc3f 100644
--- a/src/bech32.h
+++ b/src/bech32.h
@@ -12,6 +12,7 @@
 #ifndef BITCOIN_BECH32_H
 #define BITCOIN_BECH32_H
 
+#include <optional>
 #include <stdint.h>
 #include <string>
 #include <vector>
@@ -20,8 +21,6 @@ namespace bech32
 {
 
 enum class Encoding {
-    INVALID, //!< Failed decoding
-
     BECH32,  //!< Bech32 encoding as defined in BIP173
     BECH32M, //!< Bech32m encoding as defined in BIP350
 };
@@ -32,16 +31,15 @@ std::string Encode(Encoding encoding, const std::string& hrp, const std::vector<
 
 struct DecodeResult
 {
-    Encoding encoding;         //!< What encoding was detected in the result; Encoding::INVALID if failed.
+    Encoding encoding;         //!< What encoding was detected in the result
     std::string hrp;           //!< The human readable part
     std::vector<uint8_t> data; //!< The payload (excluding checksum)
 
-    DecodeResult() : encoding(Encoding::INVALID) {}
     DecodeResult(Encoding enc, std::string&& h, std::vector<uint8_t>&& d) : encoding(enc), hrp(std::move(h)), data(std::move(d)) {}
 };
 
 /** Decode a Bech32 string. */
-DecodeResult Decode(const std::string& str);
+std::optional<DecodeResult> Decode(const std::string& str);
 
 } // namespace bech32
 
diff --git a/src/key_io.cpp b/src/key_io.cpp
index dbcbfa1f29..dbb4711d42 100644
--- a/src/key_io.cpp
+++ b/src/key_io.cpp
@@ -96,25 +96,25 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
     }
     data.clear();
     const auto dec = bech32::Decode(str);
-    if ((dec.encoding == bech32::Encoding::BECH32 || dec.encoding == bech32::Encoding::BECH32M) && dec.data.size() > 0) {
+    if (dec && dec->data.size() > 0) {
         // Bech32 decoding
         error_str = "";
-        if (dec.hrp != params.Bech32HRP()) {
+        if (dec->hrp != params.Bech32HRP()) {
             error_str = "Invalid prefix for Bech32 address";
             return CNoDestination();
         }
-        int version = dec.data[0]; // The first 5 bit symbol is the witness version (0-16)
-        if (version == 0 && dec.encoding != bech32::Encoding::BECH32) {
+        int version = dec->data[0]; // The first 5 bit symbol is the witness version (0-16)
+        if (version == 0 && dec->encoding != bech32::Encoding::BECH32) {
             error_str = "Version 0 witness address must use Bech32 checksum";
             return CNoDestination();
         }
-        if (version != 0 && dec.encoding != bech32::Encoding::BECH32M) {
+        if (version != 0 && dec->encoding != bech32::Encoding::BECH32M) {
             error_str = "Version 1+ witness address must use Bech32m checksum";
             return CNoDestination();
         }
         // The rest of the symbols are converted witness program bytes.
-        data.reserve(((dec.data.size() - 1) * 5) / 8);
-        if (ConvertBits<5, 8, false>([&](unsigned char c) { data.push_back(c); }, dec.data.begin() + 1, dec.data.end())) {
+        data.reserve(((dec->data.size() - 1) * 5) / 8);
+        if (ConvertBits<5, 8, false>([&](unsigned char c) { data.push_back(c); }, dec->data.begin() + 1, dec->data.end())) {
             if (version == 0) {
                 {
                     WitnessV0KeyHash keyid;
diff --git a/src/test/bech32_tests.cpp b/src/test/bech32_tests.cpp
index 2651e46430..835d3a1c79 100644
--- a/src/test/bech32_tests.cpp
+++ b/src/test/bech32_tests.cpp
@@ -23,8 +23,9 @@ BOOST_AUTO_TEST_CASE(bech32_testvectors_valid)
     };
     for (const std::string& str : CASES) {
         const auto dec = bech32::Decode(str);
-        BOOST_CHECK(dec.encoding == bech32::Encoding::BECH32);
-        std::string recode = bech32::Encode(bech32::Encoding::BECH32, dec.hrp, dec.data);
+        BOOST_CHECK(dec);
+        BOOST_CHECK(dec->encoding == bech32::Encoding::BECH32);
+        std::string recode = bech32::Encode(bech32::Encoding::BECH32, dec->hrp, dec->data);
         BOOST_CHECK(!recode.empty());
         BOOST_CHECK(CaseInsensitiveEqual(str, recode));
     }
@@ -43,8 +44,9 @@ BOOST_AUTO_TEST_CASE(bech32m_testvectors_valid)
     };
     for (const std::string& str : CASES) {
         const auto dec = bech32::Decode(str);
-        BOOST_CHECK(dec.encoding == bech32::Encoding::BECH32M);
-        std::string recode = bech32::Encode(bech32::Encoding::BECH32M, dec.hrp, dec.data);
+        BOOST_CHECK(dec);
+        BOOST_CHECK(dec->encoding == bech32::Encoding::BECH32M);
+        std::string recode = bech32::Encode(bech32::Encoding::BECH32M, dec->hrp, dec->data);
         BOOST_CHECK(!recode.empty());
         BOOST_CHECK(CaseInsensitiveEqual(str, recode));
     }
@@ -70,7 +72,7 @@ BOOST_AUTO_TEST_CASE(bech32_testvectors_invalid)
     };
     for (const std::string& str : CASES) {
         const auto dec = bech32::Decode(str);
-        BOOST_CHECK(dec.encoding == bech32::Encoding::INVALID);
+        BOOST_CHECK(!dec);
     }
 }
 
@@ -94,7 +96,7 @@ BOOST_AUTO_TEST_CASE(bech32m_testvectors_invalid)
     };
     for (const std::string& str : CASES) {
         const auto dec = bech32::Decode(str);
-        BOOST_CHECK(dec.encoding == bech32::Encoding::INVALID);
+        BOOST_CHECK(!dec);
     }
 }
 
diff --git a/src/test/fuzz/bech32.cpp b/src/test/fuzz/bech32.cpp
index 15a3875828..c794fdafc3 100644
--- a/src/test/fuzz/bech32.cpp
+++ b/src/test/fuzz/bech32.cpp
@@ -17,10 +17,8 @@ FUZZ_TARGET(bech32)
 {
     const std::string random_string(buffer.begin(), buffer.end());
     const auto r1 = bech32::Decode(random_string);
-    if (r1.hrp.empty()) {
-        assert(r1.data.empty());
-    } else {
-        const std::string reencoded = bech32::Encode(r1.encoding, r1.hrp, r1.data);
+    if (r1) {
+        const std::string reencoded = bech32::Encode(r1->encoding, r1->hrp, r1->data);
         assert(CaseInsensitiveEqual(random_string, reencoded));
     }
 
@@ -33,9 +31,10 @@ FUZZ_TARGET(bech32)
             const std::string encoded = bech32::Encode(encoding, "bc", input);
             assert(!encoded.empty());
             const auto r2 = bech32::Decode(encoded);
-            assert(r2.encoding == encoding);
-            assert(r2.hrp == "bc");
-            assert(r2.data == input);
+            assert(r2);
+            assert(r2->encoding == encoding);
+            assert(r2->hrp == "bc");
+            assert(r2->data == input);
         }
     }
 }
```

</details>

Note that we've removed the following check from the fuzz tester:

```diff
-    if (r1.hrp.empty()) {
-        assert(r1.data.empty());
```

That inconsistent state can't exist where `DecodeResult` only contains a successfully decoded bech32{m} string.

EDIT: I guess you're not using std::optional so it's easier to backport to versions that don't require c++17?

Reply:
@jnewbery Yes, exactly. I considered that, but decided against it, as I prefer to have something maximally backportable.

Reply:
Thanks. Marking resolved.